Binaries/DafnyRuntime.h

```diff
@@ -277,12 +269,45 @@ inline int64 EuclideanDivision_int64(int64 a, int64 b) {
     }    
 }
 
+/*********************************************************
+ *  ARRAYS
+ *********************************************************/
+
 template <typename T>
-T* global_empty_ptr = new T[0];
+struct DafnyArray {
+  shared_ptr<vector<T>> vec;
+
+  DafnyArray() { }
+  DafnyArray(size_t len) {
+    vec = make_shared<vector<T>>(len);
+  }
+
+  static DafnyArray<T> Null() { return DafnyArray<T>(); }
+  static DafnyArray<T> New(size_t len) { return DafnyArray<T>(len); }
+
+  size_t size() const { return vec->size(); }
+  T& at(uint64 idx) const { return (*vec)[idx]; }
```

Comment:
Nice.  I hadn't thought about the fact that the vec->at would be bounds-checked.